Extract functions on Command Processor

huwpascoe opened the pull request:
Using the fastest method, which turns out to be an array of c-style function pointers (after testing with std::function, etc.)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/citra-emu/citra/2943)
<!-- Reviewable:end -->

lioncash commented on src/video_core/command_map.h:
```diff
@@ -0,0 +1,526 @@
+void FogLut(const Command&);
+void ProcTexLut(const Command&);
+
+typedef void (*CommandAction)(const Command&);
```
```cpp
using CommandAction = void (*)(const Command&);
```

lioncash commented on src/video_core/command_map.h:
```diff
@@ -0,0 +1,526 @@
+typedef void (*CommandAction)(const Command&);
+void nop(const Command&) {}
+
+#define _(id) nop
```
Identifiers beginning with underscores are reserved by the standard. Also, don't use [macros](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es30-dont-use-macros-for-program-text-manipulation) for program text manipulation.

huwpascoe replied:
I was basing it on the current code which does exactly the same.
#define PICA_REG_INDEX_WORKAROUND(field_name, backup_workaround_index) (backup_workaround_index)

lioncash replied:
That exists only as a workaround to a bug in the Visual Studio 2015 compiler (said bug may already be fixed in VS 2017). It's the only reason the constants were hardcoded.

lioncash commented on src/video_core/command_processor.cpp:
```diff
@@ -32,24 +33,26 @@ namespace Pica {
 
-static int vs_float_regs_counter = 0;
-static u32 vs_uniform_write_buffer[4];
+int vs_float_regs_counter = 0;
```
Why do these variables need to be non-static? They're still only used internally.

huwpascoe replied:
Why do they need to be static?

lioncash replied:
If they're only ever used in the same translation unit, they're marked static to signify internal linkage. When applied to functions, it also enables compilers to potentially inline said functions within the translation unit.

lioncash commented on src/video_core/command_processor.cpp:
```diff
@@ -6,6 +6,7 @@
 #include <cstddef>
 #include <memory>
 #include <utility>
+#include "command_map.h"
```
This should be included as `"video_core/command_map.h"`

huwpascoe commented:
> Just saying "it's faster" alone doesn't really justify the maintenance burden IMO.

A "maintenance burden"? Compared to a massive switch statement? Don't talk BS. : P

chris062689 commented:
That still doesn't address what type of systems you've tested this on, games, before and after statistics. 
We need proof. :)

huwpascoe commented:
@chris062689 Don't worry, you can [clone this PR locally](https://stackoverflow.com/questions/14947789/github-clone-from-pull-request) to test it.

lioncash commented:
Generally the onus is on the one opening the PR to provide the performance information

huwpascoe commented:
Amendments:
```
static, static everywhere
Removed the macros although command_map.h is now a lot longer with those unknown regs expanded.
Removed changes to command_processor.h
Included new file in cmakelists.txt because that would help if it could actually compile.
```
I think there's a misunderstanding though. I don't claim this will improve performance, I don't know if it will. I was assured in discord that this refactor was the right step towards future performance increases however.

B3n30 commented:
Basically, your code shouldn't really change anything on the compiled version. The compiler should compile the switch statement as well as your function array into a branch table. So there is no `right step towards future performance increases`. But I haven't checked if our compilers really do that. That is probably something that should get checked before merging this PR.

If it is compiled to a branch table than I think that PR is an improvement in case of readability and we use function arrays in other places inside citra as well. So it's just consistency.

Subv commented:
@B3n30 MSVC won't compile the switch into a jump table because it is _very_ sparse, I don't know what the generated code looks like for this PR, but i doubt it'd be much worse than a bunch of jumps strung together

B3n30 commented on src/video_core/command_processor.cpp:
```diff
@@ -119,495 +126,403 @@ static void WriteUniformFloatReg(ShaderRegs& config, Shader::ShaderSetup& setup,
+    auto& regs = g_state.regs;
+
+    if (cmd.ID >= Regs::NUM_REGS) {
+        LOG_ERROR(HW_GPU,
```
Is this how the HW reacts, just ignore the write, or does it crash?

huwpascoe commented on src/video_core/command_map.h:
```diff
@@ -0,0 +1,810 @@
+static void ProcTexLut(const Command&);
+
+using CommandAction = void (*)(const Command&);
+static const CommandAction CommandMap[] = {
```
This might make for an extremely long .cpp though?

Subv replied:
Currently it is a static variable in a header file, every translation unit that includes this file will get its own CommandMap array, potentially bloating the generated code.

huwpascoe commented:
I can't see this getting merged, maybe too big to handle. Will rethink.
